src/protocols/bdx/BdxTransferSession.cpp

```diff
@@ -0,0 +1,873 @@
+/**
+ *    @file
+ *      Implementation for the TransferSession class.
+ *      // TODO: Support Asynchronous mode. Currently, only Synchronous mode is supported.
+ */
+
+#include <protocols/bdx/BdxTransferSession.h>
+
+#include <protocols/Protocols.h>
+#include <protocols/bdx/BdxMessages.h>
+#include <support/CodeUtils.h>
+#include <system/SystemPacketBuffer.h>
+
+namespace {
+const uint8_t kBdxVersion = 0; ///< The version of this implementation of the BDX spec
+}
+
+/**
+ * @brief
+ *   Allocate a new PacketBuffer and write data from a BDX message struct.
+ */
+template <class BdxMsgType>
+CHIP_ERROR WriteToPacketBuffer(const BdxMsgType & msgStruct, ::chip::System::PacketBufferHandle & msgBuf)
+{
+    CHIP_ERROR err     = CHIP_NO_ERROR;
+    size_t msgDataSize = msgStruct.MessageSize();
+    msgBuf             = ::chip::System::PacketBufferHandle::New(static_cast<uint16_t>(msgDataSize));
+    if (msgBuf.IsNull())
+    {
+        return CHIP_ERROR_NO_MEMORY;
+    }
+
+    ::chip::BufBound bbuf(msgBuf->Start(), msgBuf->AvailableDataLength());
+    msgStruct.WriteToBuffer(bbuf);
+    msgBuf->SetDataLength(static_cast<uint16_t>(bbuf.Needed()));
```

Comment:
will add Fit() check and AvailableSize check

Reply:
Since Monday, you can simplify this to (approximately)
```
System::PacketBufBound bbuf(msgDataSize);
if (bbuf.IsNull()) return CHIP_ERROR_NO_MEMORY;
msgStruct.WriteToBuffer(bbuf);
msgBuf = bbuf.Finalize();
if (msgBuf.IsNull()) return CHIP_ERROR_NO_MEMORY;  // or CHIP_ERROR_BUFFER_TOO_SMALL ?
```
Length and size checks and `SetDataLength` are done for you by `PacketBufBound`.

Reply:
Thanks for the tip Kevin, I'll use the new semantics.